oci/private/authn.bzl

```diff
@@ -124,15 +124,93 @@ exec "docker-credential-{}" get <<< "$1" """.format(helper_name),
 
     response = json.decode(result.stdout)
 
-    if response["Username"] == "<token>":
-        fail("Identity tokens are not supported at the moment. See: https://github.com/bazel-contrib/rules_oci/issues/129")
-
     return {
         "type": "basic",
         "login": response["Username"],
         "password": response["Secret"],
     }
 
+
+def _oauth2(rctx, realm, scope, service, secret):
+    if rctx.os.name.startswith("windows") and rctx.which("powershell"):
+        executable = "oauth2.ps1"
+        rctx.file(
+            executable,
+            content = """\
+param (
+    [string]$url,
+    [string]$service,
+    [string]$scope,
+    [string]$refresh_token
+)
+
+try {
+    $response = Invoke-RestMethod -Uri $url -Method Post -Body @{
+        grant_type = "refresh_token"
+        service = $service
+        scope = $scope
+        refresh_token = $refresh_token
+    } -ErrorAction Stop
+
+    $jsonResponse = $response | ConvertTo-Json
+    echo $jsonResponse
+} catch {
+    $ErrorMessage = $_.Exception.Message
+    Write-Error "oauth2 failed: PowerShell request failed with error: $ErrorMessage"
+    exit 1
+}
+"""
+        )
+        result = rctx.execute(["powershell", "-File", rctx.path(executable), realm, service, scope, secret])
+    elif rctx.which("curl"):
+        executable = "oauth2.sh"
+        rctx.file(
+            executable,
+            content = """\
+url=$1
+service=$2
+scope=$3
+refresh_token=$4
+
+response=$(curl --silent --show-error --fail --request POST --data "grant_type=refresh_token&service=$service&scope=$scope&refresh_token=$refresh_token" $url)
+
+if [ $? -ne 0 ]; then
+    echo "oauth2 failed: curl request failed" >&2
+    exit 1
+fi
+
+echo "$response"
+"""
+        )
+        result = rctx.execute(["bash", rctx.path(executable), realm, service, scope, secret])
+    elif rctx.which("wget"):
+        executable = "oauth2.sh"
+        rctx.file(
+            executable,
+            content = """\
+url=$1
+service=$2
+scope=$3
+refresh_token=$4
+
+response=$(wget --quiet --output-document=- --post-data "grant_type=refresh_token&service=$service&scope=$scope&refresh_token=$refresh_token" $url)
+
+if [ $? -ne 0 ]; then
+    echo "oauth2 failed: wget request failed" >&2
```

Comment:
can we print more information about why it failed?

Reply:
I removed this line altogether, the failure reason is already in the stderr. For the code that you refer to, it would look like:
```
Error in fail: oauth2 failed: 
STDOUT:

STDERR:
curl: (22) The requested URL returned error: 403
oauth2 failed: curl request failed
```

after removing this line it is:

```
Error in fail: oauth2 failed: 
STDOUT:

STDERR:
curl: (22) The requested URL returned error: 403
```